File: jmix-templates/content/project/application/src/test/java/${project_rootPath}/user/UserUiTest.java
package ${project_rootPackage}.user;

import ${project_rootPackage}.${project_classPrefix}Application;
import ${project_rootPackage}.entity.User;
import ${project_rootPackage}.view.user.UserDetailView;
import ${project_rootPackage}.view.user.UserListView;
import com.vaadin.flow.component.Component;
import io.jmix.core.DataManager;
import io.jmix.flowui.ViewNavigators;
import io.jmix.flowui.component.UiComponentUtils;
import io.jmix.flowui.component.grid.DataGrid;
import io.jmix.flowui.component.textfield.JmixPasswordField;
import io.jmix.flowui.component.textfield.TypedTextField;
import io.jmix.flowui.data.DataUnit;
import io.jmix.flowui.data.grid.ContainerDataGridItems;
import io.jmix.flowui.kit.component.button.JmixButton;
import io.jmix.flowui.testassist.FlowuiTestAssistConfiguration;
import io.jmix.flowui.testassist.UiTest;
import io.jmix.flowui.testassist.UiTestUtils;
import io.jmix.flowui.view.View;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;

import java.util.Optional;

/**
 * Sample UI integration test for the User entity.
 */
@UiTest
@SpringBootTest(classes = {${project_classPrefix}Application.class, FlowuiTestAssistConfiguration.class})
public class UserUiTest {

    @Autowired
    DataManager dataManager;

    @Autowired
    ViewNavigators viewNavigators;

    @Test
    void test_createUser() {
        // Navigate to user list view
        viewNavigators.view(UserListView.class).navigate();

        UserListView userListView = UiTestUtils.getCurrentView();

        // click "Create" button
        JmixButton createBtn = findComponent(userListView, "createBtn");
        createBtn.click();

        // Get detail view
        UserDetailView userDetailView = UiTestUtils.getCurrentView();

        // Set username and password in the fields
        TypedTextField<String> usernameField = findComponent(userDetailView, "usernameField");
        String username = "test-user-" + System.currentTimeMillis();
        usernameField.setValue(username);

        JmixPasswordField passwordField = findComponent(userDetailView, "passwordField");
        passwordField.setValue("test-passwd");

        JmixPasswordField confirmPasswordField = findComponent(userDetailView, "confirmPasswordField");
        confirmPasswordField.setValue("test-passwd");

        // Click "OK"
        JmixButton commitAndCloseBtn = findComponent(userDetailView, "saveAndCloseBtn");
        commitAndCloseBtn.click();

        // Get navigated user list view
        userListView = UiTestUtils.getCurrentView();

        // Check the created user is shown in the table
        DataGrid<User> usersDataGrid = findComponent(userListView, "usersDataGrid");

        DataUnit usersDataGridItems = usersDataGrid.getItems();
        Assertions.assertNotNull(usersDataGridItems);

        //noinspection unchecked
        ((ContainerDataGridItems<User>) usersDataGridItems).getContainer()
                .getItems().stream()
                .filter(u -> u.getUsername().equals(username))
                .findFirst()
                .orElseThrow();
    }

    @AfterEach
    void tearDown() {
        dataManager.load(User.class)
                .query("e.username like ?1", "test-user-%")
                .list()
                .forEach(u -> dataManager.remove(u));
    }

    /**
     * Returns a component defined in the screen by the component id.
     * Throws an exception if not found.
     */
    @SuppressWarnings("unchecked")
    private <T> T findComponent(View<?> view, String componentId) {
        Optional<Component> component = UiComponentUtils.findComponent(view, componentId);
        Assertions.assertTrue(component.isPresent());
        return (T) component.get();
    }
}

File: jmix-templates/content/project/application-kotlin/src/test/kotlin/${project_rootPath}/user/UserUiTest.kt
package ${project_rootPackage}.user

import ${project_rootPackage}.${project_classPrefix}Application
import ${project_rootPackage}.entity.User
import ${project_rootPackage}.view.user.UserDetailView
import ${project_rootPackage}.view.user.UserListView
import io.jmix.core.DataManager
import io.jmix.flowui.ViewNavigators
import io.jmix.flowui.component.UiComponentUtils
import io.jmix.flowui.component.grid.DataGrid
import io.jmix.flowui.component.textfield.JmixPasswordField
import io.jmix.flowui.component.textfield.TypedTextField
import io.jmix.flowui.kit.component.button.JmixButton
import io.jmix.flowui.testassist.FlowuiTestAssistConfiguration
import io.jmix.flowui.testassist.UiTest
import io.jmix.flowui.testassist.UiTestUtils
import io.jmix.flowui.view.View
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest

/**
 * Sample UI integration test for the User entity.
 */
@UiTest
@SpringBootTest(classes = [${project_classPrefix}Application::class, FlowuiTestAssistConfiguration::class])
class UserUiTest {

    @Autowired
    lateinit var dataManager: DataManager

    @Autowired
    lateinit var viewNavigators: ViewNavigators

    @Test
    fun test_createUser() {
        // Navigate to user list view
        viewNavigators.view(UserListView::class.java)
                .navigate()

        var userListView = UiTestUtils.getCurrentView<UserListView>()

        // click "Create" button
        val createBtn = findComponent<JmixButton>(userListView, "createBtn")
        createBtn.click()

        // Get detail view
        val userDetailView = UiTestUtils.getCurrentView<UserDetailView>()

        // Set username and password in the fields
        val usernameField = findComponent<TypedTextField<String>>(userDetailView, "usernameField")
        val username = "test-user-" + System.currentTimeMillis()
        usernameField.value = username

        val passwordField = findComponent<JmixPasswordField>(userDetailView, "passwordField")
        passwordField.value = "test-passwd"

        val confirmPasswordField = findComponent<JmixPasswordField>(userDetailView, "confirmPasswordField")
        confirmPasswordField.value = "test-passwd"

        // Click "OK"
        val commitAndCloseBtn = findComponent<JmixButton>(userDetailView, "saveAndCloseBtn")
        commitAndCloseBtn.click()

        // Get navigated user list view
        userListView = UiTestUtils.getCurrentView()

        // Check the created user is shown in the table
        val usersDataGrid = findComponent<DataGrid<User>>(userListView, "usersDataGrid")

        val usersDataGridItems = usersDataGrid.items
        val user = usersDataGridItems!!.items
                .find { u: User -> u.getUsername() == username }

        Assertions.assertNotNull(user)
    }

    @AfterEach
    fun tearDown() {
        dataManager.load(User::class.java)
                .query("e.username like ?1", "test-user-%")
                .list()
                .forEach { u: User? -> dataManager.remove(u) }
    }

    /**
     * Returns a component defined in the screen by the component id.
     * Throws an exception if not found.
     */
    @Suppress("UNCHECKED_CAST")
    private fun <T> findComponent(view: View<*>, componentId: String): T {
        val component = UiComponentUtils.findComponent(view, componentId)
        Assertions.assertTrue(component.isPresent)
        return component.get() as T
    }
}